File: samples/FsBulletML.Sample.MonoGame.CSharp/FsBulletMLSampleGame.cs
using System;
using System.Collections.Generic;
using System.Linq;

using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;

using FsBulletML;
using FsBulletML.MonoGame;
using Settings = FsBulletML.MonoGame.Settings;
using unit = Microsoft.FSharp.Core.Unit;

namespace FsBulletML.Sample.MonoGame.CSharp
{
    public class FsBulletMLSampleGame : Game
    {
        private static Vector2 EnemyDefaultPos = new Vector2(Settings.Enemy.X, Settings.Enemy.Y);
        private static IEnumerable<BulletmlInfo> EnemyBullets = null;

        private GraphicsDeviceManager gmanager;
        private SpriteBatch spriteBatch;
        private Texture2D EnemyTexture { get; set; }
        private Texture2D PlayerBulletTexture { get; set; }
        private Texture2D EnemyBulletTexture { get; set; }
        private Texture2D BackgroundTexture { get; set; }
        private Texture2D ParticleTexture { get; set; }
        private string BulletName { get; set; }
        private IEnemy Boss { get; set; }
        private Background Background { get; set; }
        private ParticleEmitter Emitter { get; set; }
        private SpriteFont Font { get; set; }
        private Fps Fps { get; set; }
        private KeyboardState CurrentKeyState { get; set; }
        private KeyboardState PrevKeyState { get; set; }
        private int EnemyIndex { get; set; }
        public static Player Player { get; set; }

        public FsBulletMLSampleGame()
        {
            gmanager = new GraphicsDeviceManager(this);
            //Content.RootDirectory = "Content";

            gmanager.PreferredBackBufferWidth = (int)Settings.Display.Width;
            gmanager.PreferredBackBufferHeight = (int)Settings.Display.Height;
            Player = new Player();
            Player.Init();
        }

        protected override void Initialize()
        {
            base.Initialize();
        }

        protected override void LoadContent()
        {
            base.LoadContent();
            base.Window.Title = "FsBulletML.Sample.MonoGame.CSharp";
            spriteBatch = new SpriteBatch(GraphicsDevice);
            this.Font = this.Content.Load<SpriteFont>(@"..\..\Content\font\SpriteFont2");
            FsBulletMLSampleGame.EnemyBullets = EnemyControl.Bullets();

            this.Fps = new Fps();

            var enemyInfo = GetEnemyInfo(this.EnemyIndex);
            var pos = enemyInfo.Item1;
            var bullet = enemyInfo.Item2;
            this.BulletName = bullet.Name;
            this.Boss = CreateEnemy(pos, bullet); 

            Player.Texture = this.Content.Load<Texture2D>(@"..\..\Content\Sprites\player");

            this.EnemyTexture = this.Content.Load<Texture2D>(@"..\..\Content\Sprites\enemy1");
            this.PlayerBulletTexture = this.Content.Load<Texture2D>(@"..\..\Content\Sprites\p_bullet_s");
            this.EnemyBulletTexture = this.Content.Load<Texture2D>(@"..\..\Content\Sprites\g_bullet_s");
            this.BackgroundTexture = this.Content.Load<Texture2D>(@"..\..\Content\Sprites\background");
            this.ParticleTexture = this.Content.Load<Texture2D>(@"..\..\Content\Sprites\particle");

            this.Emitter = new ParticleEmitter();
            this.Background = new Background(this.BackgroundTexture, gmanager.PreferredBackBufferHeight, 64);
        }

        protected override void UnloadContent()
        {
            base.UnloadContent();
        }

        protected override void Update(GameTime gameTime)
        {
            base.Update(gameTime);

            this.Background.Update((float)gameTime.ElapsedGameTime.TotalSeconds);
            this.Fps.Update((float)gameTime.ElapsedGameTime.TotalSeconds);

            this.PrevKeyState = this.CurrentKeyState;
            this.CurrentKeyState = Keyboard.GetState(); 

            if (this.Boss.Life == 0 || (this.IsPressed(Keys.Enter)))
            {
                Manager.RemoveAll();
                this.EnemyIndex += 1;
                var enemyInfo = GetEnemyInfo(this.EnemyIndex);
                var pos = enemyInfo.Item1;
                var bullet = enemyInfo.Item2;
                ((IBullet)this.Boss).Used = false;
                this.BulletName = bullet.Name;
                this.Boss = CreateEnemy(pos, bullet);
            }

            if (Keyboard.GetState().IsKeyDown(Keys.Escape))
              this.Exit();

            Player.Update();
            Manager.Update();
            Manager.Free();
            Manager.UpdateSpace();

            Action cont1 = () => { CreateParticle(Player.Pos); Player.DamageCounter += 1; };
            Action cont2 = () => { CreateParticle(this.Boss.Pos); if (this.Boss.Life > 0) this.Boss.Life -= 1; };

            Manager.CheckPlayerCollision(Player.Pos, Player.Radius, cont1.ToFSharpFunc());
            Manager.CheckEnemyCollision(this.Boss.Pos, this.Boss.Radius, cont2.ToFSharpFunc());
            this.Emitter.Update((float)gameTime.ElapsedGameTime.TotalSeconds);
        }

        private void CreateParticle(Vector2 pos)
        {
            var rand = new Random ();
            var v = new Vector2 (rand.Next(0, 10), rand.Next(0, 10));
            this.Emitter.Emmit(this.ParticleTexture, pos + v, 1.4f, 1.2f, 1.0f, 2, 45, Color.Gold);
            this.Emitter.Emmit(this.ParticleTexture, pos + v, 1.2f, 1.5f, 1.2f, 1, 40, Color.OrangeRed);
            this.Emitter.Emmit(this.ParticleTexture, pos + v, 1.2f, 1.2f, 0.9f, 1, 45, Color.LightGoldenrodYellow);
            this.Emitter.Emmit(this.ParticleTexture, pos + v, 1.6f, 2.0f, 1.8f, 2, 40, Color.Orange);
            this.Emitter.Emmit(this.ParticleTexture, pos + v, 1.0f, 1.5f, 1.5f, 2, 40, Color.DarkRed);
            this.Emitter.Emmit(this.ParticleTexture, pos + v, 1.3f, 1.2f, 1.4f, 2, 40, Color.Tomato);
        }

        protected override void Draw(GameTime gameTime)
        {
            base.Draw(gameTime);
            spriteBatch.Begin();
            GraphicsDevice.Clear(Color.Black);
            this.Background.Draw(spriteBatch);

            DrawText ((String.Format("FPS: {0:F5}", this.Fps.Value)), new Vector2(3, 3), Color.White);

            foreach (var bullet in Manager.PlayerBullets)
            {
		        var  textureCenter = new Vector2(this.PlayerBulletTexture.Width / 2, this.PlayerBulletTexture.Height / 2);
                var position = Manager.GetDrawPos(bullet.Pos, this.PlayerBulletTexture) + textureCenter;
                spriteBatch.Draw(this.PlayerBulletTexture, position, null, Color.White,bullet.Dir, textureCenter, 1.0f, SpriteEffects.None, 0);
            }

            foreach (var bullet in Manager.EnemyBullets)
            {
                var textureCenter = new Vector2(this.EnemyBulletTexture.Width / 2, this.EnemyBulletTexture.Height / 2);
                var position = Manager.GetDrawPos(bullet.Pos, this.EnemyBulletTexture) + textureCenter;
                spriteBatch.Draw(this.EnemyBulletTexture, position, null, Color.White, bullet.Dir, textureCenter, 1.0f, SpriteEffects.None, 0);
            }

            foreach (var enemy in Manager.Enemys)
            {
                spriteBatch.Draw(this.EnemyTexture, Manager.GetDrawPos(enemy.Pos, this.EnemyTexture), Color.AntiqueWhite);
            }

            spriteBatch.Draw(Player.Texture, Manager.GetDrawPos(Player.Pos, Player.Texture), Color.AntiqueWhite);

            DrawText(string.Format("Name :{0} ",this.BulletName), new Vector2(3, 18), Color.White);
            DrawText(string.Format("Boss Life : {0}", this.Boss.Life.ToString()), new Vector2(3, 33), Color.White);
            DrawText(string.Format("Player Damage : {0}", Player.DamageCounter.ToString()), new Vector2(3, 48), Color.White);

            this.Emitter.Draw(spriteBatch);
            spriteBatch.End();
        }

        private bool IsPressed(Keys key)
        {
            return this.CurrentKeyState.IsKeyDown(key) && this.PrevKeyState.IsKeyUp(key);
        }

        private Enemy CreateEnemy(Vector2 pos, BulletmlInfo bulletMove) 
        {
            var enemy = new Enemy();
            Manager.AddEnemy(enemy);
            enemy.pos = pos;
            var name = bulletMove.Name;
            enemy.SetBulletTask(name, bulletMove.BulletmlTaskOption);
            return enemy;
        }

        private Tuple<Vector2, BulletmlInfo> GetEnemyInfo(int index) 
        {
            var len = FsBulletMLSampleGame.EnemyBullets.Count();
            if (len <= (index))
                this.EnemyIndex = 0;
            var bullet = FsBulletMLSampleGame.EnemyBullets.ElementAt(this.EnemyIndex);
            return new Tuple<Vector2, BulletmlInfo>(EnemyDefaultPos, bullet);
        }

        private void DrawText(string msg, Vector2 v, Color c)
        {
            spriteBatch.DrawString(this.Font, msg, new Vector2(v.X + 2, v.Y + 2), Color.Gray);
            spriteBatch.DrawString(this.Font, msg, v, c);
        }
    }
}

File: samples/FsBulletML.Sample.MonoGame.FSharp/FsBulletMLSampleGame.fs
namespace FsBulletML.Sample.MonoGame.FSharp

open System
open System.Collections.Generic
open System.Runtime.Serialization
open Microsoft.FSharp.Core.Operators.Unchecked 
open Microsoft.Xna.Framework
open Microsoft.Xna.Framework.Input
open Microsoft.Xna.Framework.Graphics

open FsBulletML
open FsBulletML.MonoGame

type FsBulletMLSampleGame () as this = 
  inherit Game()
  [<DefaultValue>]val mutable boss : IEnemy
  [<DefaultValue>]val mutable currentKeyState : KeyboardState 
  [<DefaultValue>]val mutable prevKeyState : KeyboardState 
  [<DefaultValue>]val mutable enemyIndex : int 
  [<DefaultValue>]val mutable background : Background
  [<DefaultValue>]val mutable fps : Fps
  [<DefaultValue>]val mutable bulletName : string
  [<DefaultValue>]val mutable enemyBullets : BulletmlInfo list
  [<DefaultValue>]val mutable emitter : ParticleEmitter

  static let mutable gmanager = null : GraphicsDeviceManager
  let gametitle, sprite = "FsBulletML.Sample.MonoGame.FSharp", lazy new SpriteBatch(this.GraphicsDevice)
  let bulletTexture,enemyBullet1Texture, enemyBullet2Texture, playerTexture, playerBullet1Texture, playerBullet2Texture, enemyTexture, backgroundTexture, particleTexture, enemyBullet3Texture = 
    ["bullet"; "enemy_bullet1";"enemy_bullet2";"player";"p_bullet_s";"player_bullet2";"enemy1";"background";"particle";"g_bullet_s"] 
    |> List.map (fun name -> lazy this.Content.Load<Texture2D>(@"..\..\..\Content\Sprites\" + name)) |> function 
    |  a::b::c::d::e::f::g::h::i::j::[] -> a,b,c,d,e,f,g,h,i,j | _ -> invalidArg "tlist" "長さが違う"
  let sfont = lazy this.Content.Load<SpriteFont>(@"..\..\..\Content\font\SpriteFont2")
  let drawText (msg:string) (v:Vector2) c = sprite.Force() |> function
    | x -> sfont.Force() |> fun font -> [font, msg, Vector2(v.X+2.f,v.Y+2.f), Color.Gray; font, msg, v, c ] 
                                       |> List.iter (fun (font, msg, v, c) -> x.DrawString(font, msg, v, c))

  static let mutable ship = defaultof<Player>
  static member Player : Player = ship
  static member Graphics = gmanager

  let enemyDefaultPos = new Vector2(Settings.Enemy.X, Settings.Enemy.Y)
  let createEnemy pos bulletMove = 
      let enemy = new Enemy()
      Manager.addEnemy(enemy)
      enemy.pos <- pos 
      enemy.SetBulletTask(bulletMove)
      enemy

  let getEnemyInfo = fun index ->
    let len = this.enemyBullets |> List.length 
    if len <= (index) then
      this.enemyIndex  <- 0
    
    this.enemyBullets
    |> Seq.nth(this.enemyIndex)
    |> fun bullet -> bullet.BulletmlTask().Init(); enemyDefaultPos, (bullet.Name, bullet)

  do 
    gmanager <- new GraphicsDeviceManager(this)
    gmanager.PreferredBackBufferWidth <- Settings.Display.Width |> int
    gmanager.PreferredBackBufferHeight <- Settings.Display.Height |> int

  override this.Initialize() =
    base.Initialize()

  override this.LoadContent() =
    base.Window.Title <-gametitle
    this.enemyBullets <- EnemyControl.bullets 

    let pos, (name,bm) = getEnemyInfo(this.enemyIndex)
    this.bulletName <- name
    this.boss <- createEnemy pos (name,bm) 

    ship <- new Player()
    ship.Init()
    ship.texture <- playerTexture.Force()

    this.emitter <- new ParticleEmitter()
    this.fps <- new Fps()
    let bgTexture = backgroundTexture.Force()
    this.background <- new Background(bgTexture, gmanager.PreferredBackBufferHeight |> float32, 64.f)
    base.LoadContent()

  override this.UnloadContent () = 
    base.UnloadContent()

  override this.Update(gameTime) = 
    this.background.Update(gameTime.ElapsedGameTime.TotalSeconds)
    this.fps.Update(gameTime.ElapsedGameTime.TotalSeconds)

    this.prevKeyState <- this.currentKeyState;
    this.currentKeyState <- Keyboard.GetState();

    if this.boss.Life = 0 || (this.IsPressed(Keys.Enter)) then
      Manager.removeAll()
      this.enemyIndex <- this.enemyIndex + 1
      let pos, (name,bm) = getEnemyInfo(this.enemyIndex)
      (this.boss :> IBullet).Used <- false
      this.bulletName <- name
      this.boss <- createEnemy pos (name,bm) 

    if Keyboard.GetState().IsKeyDown(Keys.Escape) then
      this.Exit()

    ship.Update()
    Manager.update()
    Manager.free()
    Manager.updateSpace()

    let createParticle pos =
      let rand = new Random ()
      let v = new Vector2 (rand.Next(0, 10) |> float32, rand.Next(0, 10) |> float32)
      let particleTexture = particleTexture.Force()
      this.emitter.Emmit(particleTexture, pos + v, 1.4f, 1.2f, 1.0f, 2, 45, Color.Gold);
      this.emitter.Emmit(particleTexture, pos + v, 1.2f, 1.5f, 1.2f, 1, 40, Color.OrangeRed );
      this.emitter.Emmit(particleTexture, pos + v, 1.2f, 1.2f, 0.9f, 1, 45, Color.LightGoldenrodYellow);
      this.emitter.Emmit(particleTexture, pos + v, 1.6f, 2.0f, 1.8f, 2, 40, Color.Orange);
      this.emitter.Emmit(particleTexture, pos + v, 1.0f, 1.5f, 1.5f, 2, 40, Color.DarkRed);
      this.emitter.Emmit(particleTexture, pos + v, 1.3f, 1.2f, 1.4f, 2, 40, Color.Tomato);

    Manager.checkPlayerCollision ship.Pos ship.Radius (fun () -> createParticle ship.Pos ; ship.damageCounter <- ship.damageCounter + 1)
    Manager.checkEnemyCollision this.boss.Pos this.boss.Radius (fun () -> createParticle this.boss.Pos ;if this.boss.Life > 0 then this.boss.Life <- this.boss.Life - 1)
    this.emitter.Update(gameTime.ElapsedGameTime.TotalSeconds);
    base.Update gameTime

  override this.Draw(gameTime) = base.Draw gameTime |> fun _ ->
    gmanager.GraphicsDevice.Clear(Color.Black)
    let spriteBatch = sprite.Force()
    spriteBatch.Begin()
    this.background.Draw(spriteBatch)
    drawText (String.Format("FPS: {0:F5}", this.fps.value)) (new Vector2(3.f, 3.f)) Color.White

    let enemyBullet1Texture = enemyBullet3Texture.Force()
    let enemyBullet2Texture = enemyBullet2Texture.Force()
    let playerBullet1Texture = playerBullet1Texture.Force()
    let playerBullet2Texture = playerBullet2Texture.Force()
    let enemyTexture = enemyTexture.Force()

    Manager.playerBullets
    |> Seq.iter (fun bullet -> let textureCenter = new Vector2(enemyBullet1Texture.Width / 2 |> float32, enemyBullet1Texture.Height / 2 |> float32);
                               let position = Manager.getDrawPos bullet.Pos playerBullet1Texture + textureCenter
                               spriteBatch.Draw(playerBullet1Texture, position, System.Nullable(), Color.White,bullet.Dir, textureCenter, 1.0f, SpriteEffects.None, 0.f))
    Manager.enemyBullets
    |> Seq.iter (fun bullet -> let textureCenter = new Vector2(enemyBullet1Texture.Width / 2 |> float32, enemyBullet1Texture.Height / 2 |> float32);
                               let position = Manager.getDrawPos bullet.Pos enemyBullet1Texture + textureCenter
                               spriteBatch.Draw(enemyBullet1Texture, position, System.Nullable(), Color.White,bullet.Dir, textureCenter, 1.0f, SpriteEffects.None, 0.f))
    Manager.enemys 
    |> Seq.iter (fun enemy -> spriteBatch.Draw(enemyTexture, Manager.getDrawPos enemy.Pos enemyTexture, Color.AntiqueWhite))

    spriteBatch.Draw(ship.texture, Manager.getDrawPos ship.Pos ship.texture, Color.AntiqueWhite)

    drawText (sprintf "Name :%s " this.bulletName ) (new Vector2(3.f, 18.f)) Color.White  
    drawText (sprintf "Boss Life : %s" <| this.boss.Life.ToString()) (new Vector2(3.f, 33.f)) Color.White  
    drawText (sprintf "Player Damage : %s" <| ship.damageCounter.ToString()) (new Vector2(3.f, 48.f)) Color.White  

    this.emitter.Draw(spriteBatch)

    spriteBatch.End()

  override this.EndRun () = base.EndRun()

  member this.IsPressed(key:Keys) =
    (this.currentKeyState.IsKeyDown(key) && this.prevKeyState.IsKeyUp(key));

